contracts/exchange-forwarder/contracts/src/MixinExchangeWrapper.sol

```diff
@@ -77,183 +80,243 @@ contract MixinExchangeWrapper is
+            );
+        } else {
+            LibRichErrors.rrevert(LibForwarderRichErrors.UnsupportedFeeError(order.takerFeeAssetData));
+        }
+    }
+
     /// @dev Synchronously executes multiple calls of fillOrder until total amount of WETH has been sold by taker.
     ///      Returns false if the transaction would otherwise revert.
     /// @param orders Array of order specifications.
     /// @param wethSellAmount Desired amount of WETH to sell.
     /// @param signatures Proofs that orders have been signed by makers.
-    /// @return Amounts filled and fees paid by makers and taker.
+    /// @return Amounts of WETH spent and makerAsset acquired by the taker.
     function _marketSellWeth(
         LibOrder.Order[] memory orders,
         uint256 wethSellAmount,
         bytes[] memory signatures
     )
         internal
-        returns (FillResults memory totalFillResults)
+        returns (
+            uint256 wethSpentAmount,
+            uint256 makerAssetAcquiredAmount
+        )
     {
-        bytes memory makerAssetData = orders[0].makerAssetData;
-        bytes memory wethAssetData = WETH_ASSET_DATA;
-
         uint256 ordersLength = orders.length;
-        for (uint256 i = 0; i != ordersLength; i++) {
 
-            // We assume that asset being bought by taker is the same for each order.
-            // We assume that asset being sold by taker is WETH for each order.
-            orders[i].makerAssetData = makerAssetData;
-            orders[i].takerAssetData = wethAssetData;
+        for (uint256 i = 0; i != ordersLength; i++) {
+            if (!orders[i].makerAssetData.equals(orders[0].makerAssetData)) {
+                LibRichErrors.rrevert(LibForwarderRichErrors.MakerAssetMismatchError(
+                    orders[0].makerAssetData,
+                    orders[i].makerAssetData
+                ));
+            }
 
-            // Calculate the remaining amount of WETH to sell
-            uint256 remainingTakerAssetFillAmount = _safeSub(wethSellAmount, totalFillResults.takerAssetFilledAmount);
+            // Preemptively skip to avoid division by zero in _marketSellSingleOrder
+            if (orders[i].makerAssetAmount == 0 || orders[i].takerAssetAmount == 0) {
+                continue;
+            }
 
-            // Attempt to sell the remaining amount of WETH
-            FillResults memory singleFillResults = _fillOrderNoThrow(
+            LibFillResults.FillResults memory singleFillResults = _marketSellSingleOrder(
                 orders[i],
-                remainingTakerAssetFillAmount,
-                signatures[i]
+                signatures[i],
+                wethSellAmount,
+                wethSpentAmount
             );
 
-            // Update amounts filled and fees paid by maker and taker
-            _addFillResults(totalFillResults, singleFillResults);
+            // Percentage fee
+            if (orders[i].takerFeeAssetData.equals(orders[i].makerAssetData)) {
```

Comment:
Rather than repeating these `if...else` blocks, can we put this logic into `_marketSellSingleOrder`? We should already have all of the information we need there. It can probably just return `makerAssetAcquiredAmount` and `wethSpentAmount` rather than `singleFillResults`.